### PR TITLE
17 refactoring test execution system

### PR DESCRIPTION
**Pull request by ohmushi**
Refacto all the execution system, mainly by Inversion Of Control : now we pass the "compiler" and the "runner" in params (and a "infrastructure" struct containing function pointers for fopen, popen and pclose).
It's so much verbose :/ (sorry 🙏 )
But...
It is very testable and quite flexible I believe.

The compiler and the runner return a kata_run_result_t containing satus & output.
The runner takes the result of the compilation.

In the tests, there is still some stuff which is not windows friendly (memory_fopen), but I opened an issue (https://github.com/c-lings/clings/issues/18) to make it all "windows friendly" which will not be that easy to do 😞

### CODE REVIEW

File: src/configuration/get_kata_list/in_memory_kata_list.c [hunk 1]
```diff
@@ -7,6 +7,8 @@ typedef struct kata_file {
         const char* directory;
 } kata_file_t;
 
+#define MAX_LINE_SIZE 64
```
Comment (Trevodorax) [inline src/configuration/get_kata_list/in_memory_kata_list.c:10]: Isn't that a bit short ?

File: src/result_handler/result_handler.c [hunk 2]
```diff
@@ -1,29 +1,22 @@
 #include <stdio.h>
 #include "result_handler.h"
+#include "stdbool.h"
 
-void handle_kata_result(kata_status result_status, sized_string_t result_string) {
-    switch (result_status) {
-        case KATA_SUCCESS:
-            printf("Success !");
-            break;
+bool isStatusError(kata_status status) {
```
Comment (Trevodorax) [inline src/result_handler/result_handler.c:5]: Why is that one camelCase ?

File: src/types/mocks.c [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    return NULL;
+}
+
+fopen_f failing_fopen(void) {
+    return &failing_str_str;
+}
+
+FILE * memory_fopen_f(__attribute__((unused)) const char * filename, const char * mode) {
+    // create a FILE * from a buffer
+    // POSIX => not Windows friendly
```
Comment (Trevodorax) [inline src/types/mocks.c:30]: All my homies hate Windows

File: src/result_handler/result_handler.c (hunk 2 above)
Comment (ohmushi) [reply src/result_handler/result_handler.c:5]: oops

File: src/configuration/get_kata_list/in_memory_kata_list.c (hunk 1 above)
Comment (ohmushi) [reply src/configuration/get_kata_list/in_memory_kata_list.c:10]: Indeed, I will push it to 256

File: src/types/mocks.c (hunk 3 above)
Comment (ohmushi) [reply src/types/mocks.c:30]: Oh dear...we’re gonna make enemies ...

**Review by Trevodorax — APPROVED** [review]

**Pull request merged**